ProcessTree.h

```diff
@@ -0,0 +1,144 @@
+public:
+    ProcessTreeItem(enum ProcessTreeSource source, int pid, int ppid=0):
+        _source(source), _pid(pid), _ppid(ppid), _uid(-1), _gid(-1), _flags(0), _exec_propagation(0), _exited(false) {}
+    ProcessTreeItem(enum ProcessTreeSource source, int pid, int ppid, int uid, int gid, const std::string& exe, const std::string& cmdline):
+        _source(source), _pid(pid), _ppid(ppid), _uid(uid), _gid(gid), _exe(exe), _cmdline(cmdline),
+        _flags(0), _exec_propagation(0), _exited(false) {}
+
+    enum ProcessTreeSource _source;
+    int _pid;
+    int _ppid;
+    int _uid;
+    int _gid;
+    std::vector<int> _children;
+    std::vector<struct Ancestor> _ancestors;
+    unsigned int _exec_propagation;
+    std::string _exe;
+    std::string _cmdline;
+    std::bitset<FILTER_BITSET_SIZE> _flags;
+    bool _exited;
+    std::chrono::system_clock::time_point _exit_time;
+};
+
+class ProcessTree;
+
+// Class that monitors pnotify events and writes them to ProcessTree queues
+class ProcessNotify: public RunBase {
+public:
+    ProcessNotify(std::shared_ptr<ProcessTree> processTree): _processTree(processTree) {
+        InitProcSocket();
+    }
+
+protected:
+    void run() override;
+
+private:
+    void InitProcSocket();
+
+    std::shared_ptr<ProcessTree> _processTree;
+    int _proc_socket;
+};
+
+enum ProcessQueueType { ProcessQueueFork, ProcessQueueExec, ProcessQueueExit };
+
+struct ProcessQueueItem {
+    enum ProcessQueueType type;
+    enum ProcessTreeSource source;
+    int pid;
+    int ppid;
+};
+
+// Class that manages the process tree
+class ProcessTree: public RunBase {
+public:
+    ProcessTree(const std::shared_ptr<UserDB>& user_db, std::shared_ptr<FiltersEngine> filtersEngine): _user_db(user_db), _filtersEngine(filtersEngine), _queue_data_ready(false)
+    {
+        _last_clean_time = std::chrono::system_clock::now();
+    }
+
+    void AddPnForkQueue(int pid, int ppid);
+    void AddPnExecQueue(int pid);
+    void AddPnExitQueue(int pid);
+    std::shared_ptr<ProcessTreeItem> AddProcess(int pid, int ppid, int uid, int gid, std::string exe, const std::string& cmdline);
+    void Clean();
+    std::shared_ptr<ProcessTreeItem> GetInfoForPid(int pid);
+    void PopulateTree();
+    void ShowTree();
+    void ShowProcess(std::shared_ptr<ProcessTreeItem> p);
+
+protected:
+    void run() override;
+
+private:
+    void AddPid(int pid, int ppid);
+    void AddPid(int pid);
+    void RemovePid(int pid);
+    std::string ReadFirstLine(const std::string& file);
+    std::string ReadParam(const std::string& file, const std::string& param);
+    std::shared_ptr<ProcessTreeItem> ReadProcEntry(int pid);
+    bool is_number(char *s);
+    void ApplyFlags(std::shared_ptr<ProcessTreeItem> process);
+
+    std::shared_ptr<UserDB> _user_db;
+    std::shared_ptr<FiltersEngine> _filtersEngine;
+    std::unordered_map<int, std::shared_ptr<ProcessTreeItem>> _processes;
+    bool _queue_data_ready;
+    std::mutex _queue_pop_mutex;
```

Comment:
std::queue does not support concurrent push and pop operations so having separate push and pop locks is still racy.

Reply:
Thanks - changed to one lock.